scheds/rust/scx_tickless/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,696 @@
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_ctx);
+} task_ctx_stor SEC(".maps");
+
+/*
+ * Return the context of @p, or NULL if the task doesn't exist.
+ */
+struct task_ctx *try_lookup_task_ctx(const struct task_struct *p)
+{
+	return bpf_task_storage_get(&task_ctx_stor,
+				    (struct task_struct *)p, 0, 0);
+}
+
+/*
+ * Return true if the target task @p is a kernel thread.
+ */
+static inline bool is_kthread(const struct task_struct *p)
+{
+	return p->flags & PF_KTHREAD;
+}
+
+/*
+ * Keep track of the current global vruntime.
+ *
+ * The vuntime is defined as the task's runtime inversely scaled using the
+ * task's weight (priority).
+ */
+static u64 vtime_now;
+
+/*
+ * Scale value inversely proportional to the weight of @p.
+ */
+static u64 scale_inverse_fair(const struct task_struct *p, u64 value)
+{
+	return value * 100 / p->scx.weight;
+}
+
+/*
+ * Update and return the task's deadline.
+ */
+static u64 task_deadline(const struct task_struct *p, struct task_ctx *tctx)
+{
+	u64 vtime_min;
+
+	/*
+	 * Limit the amount of vtime budget that an idling task can
+	 * accumulate to prevent excessive prioritization of sleeping
+	 * tasks.
+	 *
+	 * Tasks with a higher weight get a bigger "bucket" for their
+	 * allowed accumulated time budget.
+	 */
+	vtime_min = vtime_now - slice_ns;
+	if (time_before(tctx->deadline, vtime_min))
+		tctx->deadline = vtime_min;
+
+	/*
+	 * Add the execution vruntime to the deadline.
+	 */
+	tctx->deadline += scale_inverse_fair(p, tctx->exec_runtime);
+
+	return tctx->deadline;
+}
+
+/*
+ * Return the time interval between two ticks in ns.
+ */
+static inline u64 tick_interval_ns(void)
+{
+	u64 freq = tick_freq ? : CONFIG_HZ;
+
+	return NSEC_PER_SEC / freq;
+}
+
+s32 BPF_STRUCT_OPS(tickless_select_cpu, struct task_struct *p, s32 prev_cpu, u64 wake_flags)
+{
+	/*
+	 * Always route wakeups to a CPU from the primary group to minimize
+	 * noise on the other CPUs.
+	 */
+	return pick_primary_cpu();
+}
+
+void BPF_STRUCT_OPS(tickless_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	struct task_ctx *tctx;
+	u64 deadline;
+
+	/*
+	 * Insert the task to the shared queue.
+	 */
+	tctx = try_lookup_task_ctx(p);
+	if (!tctx)
+		return;
+	deadline = task_deadline(p, tctx);
+	scx_bpf_dsq_insert_vtime(p, SHARED_DSQ, SCX_SLICE_INF, deadline, enq_flags);
+
+	/*
+	 * Trigger a resched event on a primary CPU when a task that was not
+	 * already running is enqueued.
+	 */
+	if (!scx_bpf_task_running(p))
+		scx_bpf_kick_cpu(pick_primary_cpu(), SCX_KICK_PREEMPT);
+}
+
+/*
+ * Try to consume a task from the shared queue and dispatch on a target
+ * @cpu. If @same_cpu is true, try to consume a task that was previously
+ * running on @cpu.
+ *
+ * Return true if a task was dispatched, false otherwise.
+ */
+static bool dispatch_cpu(s32 cpu, bool same_cpu)
+{
+	struct task_struct *p;
+	bool dispatched = false;
+
+	bpf_for_each(scx_dsq, p, SHARED_DSQ, 0) {
+		p = bpf_task_from_pid(p->pid);
```

Comment:
Maybe add a comment explaining that this is working around a BPF issue and isn't needed on newer kernels?